crates/smelt-graph/src/executor/remote.rs

```diff
@@ -0,0 +1,192 @@
+
+use crate::executor::Executor;
+use crate::Command;
+
+use super::common::{create_test_result, prepare_workspace, Workspace};
+
+type TRMap = Arc<HashMap<String, tokio::sync::oneshot::Sender<TestResult>>>;
+
+/// This is a dummy executor to test all of the logic of the slurm executor, with none of the
+/// overhead of creating a slurm cluster
+pub struct RemoteExecutor {
+    binary_path: PathBuf,
+}
+
+#[derive(Debug, Clone)]
+struct RemoteServer {
+    tx_chan: Sender<Event>,
+    connections: Arc<HashMap<String, tokio::sync::oneshot::Sender<TestResult>>>,
+}
+
+const WORKER_BIN: &[u8] = include_bytes!(env!("CARGO_BIN_FILE_SMELT_SLURM_worker"));
+
+async fn make_temp_executable(data: &[u8]) -> anyhow::Result<PathBuf> {
+    let file = PathBuf::from(format!(
+        "{}/workerguy",
+        std::env!("CARGO_MANIFEST_DIR").to_string()
+    ));
+
+    tokio::fs::write(file.as_path(), data).await?;
+    let mut perms = tokio::fs::metadata(file.as_path()).await?.permissions();
+    perms.set_mode(0o755); // make exec
+    tokio::fs::set_permissions(file.as_path(), perms).await?;
+    Ok(file)
+}
+
+struct PerTxRemoteState {
+    connections: TRMap,
+    server_addr: SocketAddr,
+    server_handle: JoinHandle<()>,
+}
+
+impl RemoteExecutor {
+    pub async fn new() -> Self {
+        let res = make_temp_executable(WORKER_BIN).await.unwrap();
+        Self { binary_path: res }
+    }
+}
+
+#[tonic::async_trait]
+impl smelt_data::event_listener_server::EventListener for RemoteServer {
+    async fn send_event(
+        &self,
+        request: tonic::Request<Event>,
+    ) -> std::result::Result<tonic::Response<()>, tonic::Status> {
+        let inner_event = request.into_inner();
+
+        let _resp = self.tx_chan.send(inner_event).await;
+        Ok(Response::new(()))
+    }
+    async fn send_outputs(
+        &self,
+        request: tonic::Request<TestResult>,
+    ) -> std::result::Result<tonic::Response<()>, tonic::Status> {
+        let val = request.into_inner();
+        let v = self.connections.remove(&val.test_name);
+        match v {
+            None => {
+                tracing::error!("Missing entry in the remote server!");
+                panic!();
+            }
+            Some(entry) => {
+                let _ = entry.1.send(val);
+            }
+        };
+        Ok(Response::new(()))
+    }
+}
+
+trait RemoteHelpers {
+    fn set_pertx_state(&mut self, pertxstate: PerTxRemoteState);
+    fn get_pertx_state(&self) -> Arc<PerTxRemoteState>;
+}
+
+impl RemoteHelpers for UserComputationData {
+    fn set_pertx_state(&mut self, map: PerTxRemoteState) {
+        self.data.set(Arc::new(map));
+    }
+    fn get_pertx_state(&self) -> Arc<PerTxRemoteState> {
+        self.data.get().cloned().unwrap()
+    }
+}
+
+#[async_trait]
+impl Executor for RemoteExecutor {
+    async fn init_per_tx_state(&self, data: &mut UserComputationData) {
+        // This is bad! we could collide on port! I dont care
+        let port = 9213;
```

Comment:
if we have two tasks running simultaneously we _will_ collide on port right? can we at least make sure the error message is pretty obvious?

Reply:
yeah, this will manifest if we have more than one test -- tbh, dont care to fix this right now because i'll be making some broader changes to address this with the slurm executor